File: Java/Min-Stack.java
public class Solution {
    private Stack<Integer> stk1;
    private Stack<Integer> stk2;
    public Solution() {
        // do initialize if necessary
        stk1 = new Stack<Integer>();
        stk2 = new Stack<Integer>();
    }

    public void push(int number) {
        // write your code here
        stk1.push(number);
        if (stk2.empty() || number <= min()) {
            stk2.push(number);
        } else {
            stk2.push(min());
        }
    }

    public int pop() {
        // write your code here
        stk2.pop();
        return stk1.pop();
    }

    public int min() {
        // write your code here
        return stk2.peek();
    }
}

File: Cpp/Min-Stack.cpp
class MinStack {
private:
    stack<int> st1;
    stack<int> st2;
public:
    MinStack() {
        // do initialization if necessary
    }

    void push(int number) {
        // write your code here
        st1.push(number);
        if (st2.empty()) {
            st2.push(number);
        } else {
            if (number <= st2.top()) {
                st2.push(number);
            } else {
                st2.push(this.min())
            }
        }
    }

    int pop() {
        int cnt = st1.top();
        st1.pop();
        st2.pop();
        return cnt;
        // write your code here
    }

    int min() {
        return st2.top();
        // write your code here
    }
};
